### DISCUSSION TITLE
What is the purpose of square bracket? and when must it be used?

### DISCUSSION THREAD

**Discussion by Inc0n** [hylang/hy#2200] [Q&A] [answered]
Hi, I have been trying to get my teeth in hy, and I must admit this shell is fairly tough.

The argument list in `defn` uses [], and will complain if () is used.
```
(defn example-square [] 1)
(defn example-round () 1)
```
Is there a reason why they cannot be used interchangably (I noticed how there is no quoting in hy, is that related), is this similar to how clojure works (sorry, I am not familiar with clojure at all, so I take this chance to ask it here).

Are there any other use cases where [] must be used?

On another note, I found out that `let` is not defined in `hy 0.20.0`, `python 3.9.9`. And I couldn't quite find a way to define local bindings (in the way such as let) in the documentation, other than that `setv` would work but would be more verbose than if `let` is used.

**Comment by allison-casey** [comment, marked as answer]
So couple questions going on here. 

1. If you're just getting into Hy, it would be best to start using the alphas which you can install with `pip install --pre hy` and it's companion library [hyrule](https://github.com/hylang/hyrule) (which is basically Hy's stdlib). The alpha branch contains many new features and additions (including `let`) that we're adding in order to get to `1.0`. 
2. As such you should also be using the alpha version of the docs which you can find [here](https://docs.hylang.org/en/alpha/index.html) (the dropdown at the bottom left will let you switch between versions)
3. Hy does allow for quoting with `'`, `` ` ``, `~`, and `~@` (you can read about quoting [here](https://docs.hylang.org/en/alpha/tutorial.html#macros))
4. It's generally not idiomatic in Hy to use `'(1 2 3)` to define lists, we're lispy python and so it's more appropriate to use python lists `[1 2 3]` or tuples `(, 1 2 3)` for that. Hy prefers to use `(form ...)` to indicate some kind of expression and bracketed lists `[arg1 arg2]` to define argument lists and the like. i.e. `(with [f (open "file.txt")] ...)`, `(let [a 1 b 2] ...)`, `(try ... (except [e ValueError] ...))`, `(import pathlib [Path])` etc. The [api docs](https://docs.hylang.org/en/alpha/api.html) and the [tutorial](https://docs.hylang.org/en/alpha/tutorial.html) can give you more examples of where and when to use what. 

Hopefully that answer's your questions! Feel free to ask if you have more and have fun messing around with Hy!

**Reply by Inc0n** [reply]
Thanks this is very comprehensive!

**Comment by Kodiologist** [comment]
> and I couldn't quite find a way to define local bindings (in the way such as let) in the documentation, other than that `setv` would work but would be more verbose than if `let` is used.

In most cases, `setv` should be if anything less verbose than `let`. Do you have an example where it's more verbose?
